### PR TITLE
Refactor the placer debug code

### PR DESCRIPTION
**Pull request by MohamedElgammal**
#### Description

This PR follows #1287 . In this PR, some code refactoring was done while preserving the same functionality.
#### Motivation and Context

Taking into consideration some code review comments on #1287 

#### How Has This Been Tested?

test that the UI is working for multiple VPR benchmarks
test that the feature is disabled if the compilation or the code flag is disabled

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ x] All new and existing tests passed

### CODE REVIEW

File: vpr/src/draw/draw.cpp [hunk 1]
```diff
@@ -998,16 +998,16 @@ static void drawplace(ezgl::renderer* g) {
                 ezgl::color block_color;
                 t_logical_block_type_ptr logical_block_type = nullptr;
 #    ifdef VTR_ENABLE_DEBUG_LOGGING
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:1000]: Do we need this guard?  Does it only control the breakpoint functionality or something else?  Does compiling in breakpoints slow the tool down?
If not, we can just eliminate this #ifdef throughout the code.

Comment (vaughnbetz) [reply vpr/src/draw/draw.cpp:1000]: I'd factor this into a subroutine; shorter routines with meaningful names are easier to understand and change.

File: vpr/src/draw/draw.cpp [hunk 2]
```diff
@@ -998,16 +998,16 @@ static void drawplace(ezgl::renderer* g) {
                 ezgl::color block_color;
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:998]: Put the continue for if (bnum == ...) on a separate, indented line.

File: vpr/src/draw/draw.cpp [hunk 3]
```diff
@@ -998,16 +998,16 @@ static void drawplace(ezgl::renderer* g) {
...
-                if (f_placer_debug) {
-                    t_pl_loc curr_loc;
-                    curr_loc.x = i;
-                    curr_loc.y = j;
-                    auto it = std::find_if(draw_state->colored_blocks.begin(), draw_state->colored_blocks.end(), [&curr_loc](const std::pair<t_pl_loc, ezgl::color>& a) { return (a.first.x == curr_loc.x && a.first.y == curr_loc.y); });
-                    if (it != draw_state->colored_blocks.end()) {
+                t_pl_loc curr_loc;
+                curr_loc.x = i;
+                curr_loc.y = j;
+                auto it = std::find_if(draw_state->colored_blocks.begin(), draw_state->colored_blocks.end(), [&curr_loc](const std::pair<t_pl_loc, ezgl::color>& a) { return (a.first.x == curr_loc.x && a.first.y == curr_loc.y); });
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:1004]: Use a better name than a.
colored_blocks should really be called colored_locations (I think).

File: vpr/src/draw/draw.cpp [hunk 4]
```diff
@@ -998,16 +998,16 @@ static void drawplace(ezgl::renderer* g) {
...
+                if (placer_debug_enabled() && it != draw_state->colored_blocks.end()) {
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:1005]: Is placer_debug_enabled() the right name?  Does this really mean we are in a breakpoint?  If so worth renaming.

Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:1005]: Comment: found a colored location at the spot I am drawing (currently used for drawing the current move).  This overrides any block color.

File: vpr/src/draw/draw.cpp [hunk 5]
```diff
@@ -998,16 +998,16 @@ static void drawplace(ezgl::renderer* g) {
...
                         block_color = it->second;
                         auto tile_type = device_ctx.grid[i][j].type;
                         logical_block_type = pick_best_logical_type(tile_type);
                     } else {
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:1009]: No color specified at this location; use the block color.

File: vpr/src/draw/draw.cpp [hunk 6]
```diff
@@ -4157,4 +4135,16 @@ static void run_graphics_commands(std::string commands) {
     ++draw_state->sequence_number;
 }
 
+#ifdef VTR_ENABLE_DEBUG_LOGGING
+// location highlighting functions 
+void set_draw_loc_color(t_pl_loc loc, ezgl::color clr){
+    t_draw_state* draw_state = get_draw_state_vars();
+    draw_state->colored_blocks.push_back(std::make_pair(loc,clr));
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:4142]: change variable name to colored_locs

File: vpr/src/draw/draw.cpp [hunk 7]
```diff
@@ -4157,4 +4135,16 @@ static void run_graphics_commands(std::string commands) {
     ++draw_state->sequence_number;
 }
 
+#ifdef VTR_ENABLE_DEBUG_LOGGING
+// location highlighting functions
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:4139]: Expand this:  pass in an (x,y,subtile) location and the color in which it should be drawn.  This overrides the color of any block placed in that location, and also applies if the location is empty.

File: vpr/src/draw/draw.cpp [hunk 8]
```diff
@@ -4157,4 +4135,16 @@ static void run_graphics_commands(std::string commands) {
     ++draw_state->sequence_number;
 }
 
+#ifdef VTR_ENABLE_DEBUG_LOGGING
+// location highlighting functions 
+void set_draw_loc_color(t_pl_loc loc, ezgl::color clr){
+    t_draw_state* draw_state = get_draw_state_vars();
+    draw_state->colored_blocks.push_back(std::make_pair(loc,clr));    
+}
+void clear_colored_blocks(){
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.cpp:4144]: Add a blank line (vertical spacing).
Add a function level comment.

File: vpr/src/draw/draw.h [hunk 9]
```diff
@@ -84,6 +84,12 @@ float get_net_alpha();
 
 ezgl::color get_block_type_color(t_physical_tile_type_ptr type);
 
+#ifdef VTR_ENABLE_DEBUG_LOGGING
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.h:87]: Remove the #ifdef (unless you show a slowdown).

File: vpr/src/draw/draw.h [hunk 10]
```diff
@@ -84,6 +84,12 @@ float get_net_alpha();
...
+// location highlighting functions 
+void set_draw_loc_color(t_pl_loc , ezgl::color );
```
Comment (vaughnbetz) [inline vpr/src/draw/draw.h:89]: Move the function level comments (or duplicate them) here.

File: vpr/src/place/move_utils.cpp [hunk 11]
```diff
@@ -8,9 +8,8 @@
 //Run-time flag to control when placer debug information is printed
```
Comment (vaughnbetz) [inline vpr/src/place/move_utils.cpp:8]: First two lines of this comment can probably be deleted. (get rid of #ifdef and second line seems less detailed / accurate than the rest of the comment).

File: vpr/src/place/move_utils.h [hunk 12]
```diff
@@ -57,4 +51,8 @@ bool find_to_loc_uniform(t_logical_block_type_ptr type,
                          float rlim,
                          const t_pl_loc from,
                          t_pl_loc& to);
+
+//Accessor and setter for f_placer_debug
```
Comment (vaughnbetz) [inline vpr/src/place/move_utils.h:55]: Explain what setting this to true or false means.  (idea: person who wants to call these should only have to read the .h file.

Comment (vaughnbetz) [reply vpr/src/place/move_utils.h:55]: Is placer_debug_enabled the right name?  Or should it be placer_breakpoint_reached? (or something like that).

File: vpr/src/place/place.cpp [hunk 13]
```diff
@@ -45,13 +45,17 @@
 #include "tatum/TimingReporter.hpp"
 
 #ifdef VTR_ENABLE_DEBUG_LOGGING
```
Comment (vaughnbetz) [inline vpr/src/place/place.cpp:47]: Get rid of ifdef if possible.

File: vpr/src/place/place.cpp [hunk 14]
```diff
@@ -45,13 +45,17 @@
...
+#    include "draw.h"
 #    include "draw_types.h"
 #    include "draw_global.h"
 #    include "draw_color.h"
 #    include "breakpoint.h"
 //map of the available move types and their corresponding type number
 std::map<int, std::string> available_move_types = {
     {0, "Uniform"}};
+
+#define OLD_BLK_LOC_COLOR blk_GOLD
```
Comment (vaughnbetz) [inline vpr/src/place/place.cpp:57]: Shouldn't these colors be defined in a draw routine?

File: vpr/src/place/place.cpp [hunk 15]
```diff
@@ -3046,13 +3050,13 @@ void stop_placement_and_check_breakopints(t_pl_blocks_to_be_moved& blocks_affect
         get_bp_state_globals()->get_glob_breakpoint_state()->from_block = size_t(blocks_affected.moved_blocks[0].block_num);
 
         //check for breakpoints
-        f_place_debug = check_for_breakpoints(true);
-        if (f_place_debug)
+        set_placer_debug(check_for_breakpoints(true));
```
Comment (vaughnbetz) [inline vpr/src/place/place.cpp:3053]: Seems like you should do something like:
if (check_for_breakpoints (true)) {  // What does the true flag passed in mean?  Comment this.
   set_placer_debug (true);
   ....
else {

File: vpr/src/place/place.cpp [hunk 16]
```diff
@@ -3037,7 +3041,7 @@ void transform_blocks_affected(t_pl_blocks_to_be_moved blocksAffected) {
 }
 
 #ifdef VTR_ENABLE_DEBUG_LOGGING
```
Comment (vaughnbetz) [inline vpr/src/place/place.cpp:3043]: Minimize added code in try_swap (already a critical routine and overly long).  Create high-level functions, if possible in a new place_debug.cpp (or some such) file and call them from here.  Should only see a few calls / condition checks in try_swap for breakpoints and they should be well commented.

Comment (vaughnbetz) [reply vpr/src/place/place.cpp:3043]: Looking closer, I think Mahshad already did this.

File: vpr/src/place/place.cpp [hunk 17]
```diff
@@ -3067,6 +3071,10 @@ void stop_placement_and_check_breakopints(t_pl_blocks_to_be_moved& blocks_affect
```
Comment (vaughnbetz) [inline vpr/src/place/place.cpp:3071]: Consider if this routine should be moved to another file, in /draw or in a placer breakpoints .cpp file (can't recall the exact names).

**Comment by vaughnbetz** [comment]
Various inline comments above ... please take a crack at them and let me know when it's ready for merging.

File: vpr/src/place/place.cpp (hunk 15 above)
Comment (MohamedElgammal) [reply vpr/src/place/place.cpp]: After reviewing the code, I found that the true flag means that the function is called from placer. I added a comment by this meaning.

**Comment by MohamedElgammal** [comment]
@vaughnbetz , I have pushed a new commit that addresses all your comments. please let me know if there are any further comments.

**Comment by vaughnbetz** [comment]
Looks good.  The QoR failure was just a golden result that needed updating.  Merging.

**Pull request merged**